tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -3075,9 +3002,53 @@ void LegalizeTFL::runOnFunction() {
   DEF_PATTERN_INSERT(TFLFakeQuant);
   DEF_PATTERN_INSERT(TFLOneHot);
 
-  if (failed(applyFullConversion(func, target, std::move(patterns)))) {
+  if (failed(applyPatternsAndFoldGreedily(func, std::move(patterns)))) {
+    signalPassFailure();
+  }
+
+  // This strips all of the quant types replacing them with their integer
+  // types. It interferes with the shape inference code and must be performed
+  // separately.
+  RewritePatternSet quantPatterns(&getContext());
+  quantPatterns.insert<GenericTypeConvert>(ctx, converter);
+  populateFuncOpTypeConversionPattern(quantPatterns, converter);
+  if (failed(applyFullConversion(func, target, std::move(quantPatterns)))) {
     signalPassFailure();
   }
+
+  // Insert UnrealizedConversionCasts to guarantee ReturnOp agress with
+  // the FuncOp type.
+  IRRewriter rewriter(func.getContext());
+  func.walk([&](ReturnOp op) {
+    FuncOp parent = dyn_cast<FuncOp>(op->getParentOp());
+    if (!parent) return;
+
+    rewriter.setInsertionPoint(op);
+    FunctionType funcTy = func.getType();
+    auto resultTys = funcTy.getResults();
+
+    bool castAdded = false;
+    SmallVector<Value> castedValues;
+    for (auto it : llvm::zip(op->getOperands(), resultTys)) {
+      auto operand = std::get<0>(it);
```

Comment:
nit: Can you make these explicitly types (a couple too many levels of auto for my taste)

Reply:
Auto finished.